057-run-zk-kraft-clusters-parallel.md

```diff
@@ -0,0 +1,65 @@
+
+## Proposal
+
+The proposal is about adding a new `strimzi.io/kraft` annotation to the `Kafka` custom resource, to be applied by the user, in order to identify a ZooKeeper or KRaft based cluster.
+The possible values would be:
+
+* `disabled`, missing or any invalid value: identifies a ZooKeeper-based cluster.
+* `enabled`: identifies a KRaft-based cluster.
+
+This way, during the reconciliation, the operator is able to "detect" a ZooKeeper-based cluster avoiding the warning and allowing the user to operate it.
+On the KRaft side, the annotation would be needed to have the operator reconciling the corresponding `Kafka` custom resource.
+Without the annotation, but the `UseKRaft` feature gate enabled, the operator would try to handle it as a ZooKeeper-based one.
+The `UseKRaft` feature gate does currently not support any upgrades of KRaft clusters.
+It is expected to be used only for short-lived clusters used for development and testing.
+So no clusters are expected to exist.
+Therefore adding the annotation now does not present any backwards compatibility issues.
+Only ZooKeeper-based clusters are expected and newly created KRaft-based clusters having the `strimzi.io/kraft: enabled`.
+This approach is actually the same as for the node pools: enabling the `KafkaNodePools` feature gate on the operator is not enough and the user needs to apply the `strimzi.io/node-pools: enabled` annotation on the `Kafka` custom resource using node pools for brokers (and controllers, if KRaft is enabled as well).
+
+As a non-goal of this proposal, the same annotation could be even used for handling the ZooKeeper to KRaft migration steps as described in the proposal [PR#90](https://github.com/strimzi/proposals/pull/90).
+
+## Affected/not affected projects
+
+The Strimzi Cluster Operator is the only project affected by this proposal.
+Its logic needs to be updated in order to be able to handle the new `strimzi.io/kraft` annotation (or the missing of it).
+No other projects in the Strimzi ecosystem are impacted by this proposal.
+
+## Compatibility
+
+If the user is running the operator with `UseKRaft` feature gate enabled, it will detect ZooKeeper-based cluster allowing to operate them again.
+If the user has a KRaft-based cluster deployed, because of the missing annotation, the operator will try to handle it as it was ZooKeeper-based which would fail.
+In this case, if the user applies `strimzi.io/kraft: enabled` annotation on it, there could be unexpected results on the reconciliation. So doing this is not supported.
+In general, when this proposal is in place, the expectation is that there won't be KRaft-based clusters already running but there could be only ZooKeeper-based clusters running.
+Breaking this backward compatibility is expected by taking into account that the KRaft support in Strimzi is still behind a feature gate and should be considered just for development purposes. 
+
```

Comment:
I'd like to see mention of what happens to this annotation when

* KRaft is the only thing Kafka supports.
* An existing KRaft cluster has its annotation changed to `disabled`. 

The problem I see with it being a (user managed) annotation is that while that's fine for the user to express their intent, it cannot also be used by the operator to express how it interprets the resource.

Reply:
> I'd like to see mention of what happens to this annotation when ...

I would expect the annotation handling just removed from the operator and the annotation ignored. It's going to happen the same way for the `node-pools` annotation as stated in the corresponding proposal: "So once the ZooKeeper support is removed from Strimzi, we will also remove the support for the virtual node pool and the strimzi.io/node-pools annotation" ... details [here](https://github.com/strimzi/proposals/blob/main/050-Kafka-Node-Pools.md).

> it cannot also be used by the operator to express how it interprets the resource.

I am not sure what do you mean by this.